File: sdk/mobile/android/src/main/kotlin/com/summit/odps/android/DifferentialPrivacy.kt
package com.summit.odps.android

import kotlin.math.abs
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.sign
import kotlin.math.roundToInt
import kotlin.random.Random

/**
 * Interface for generating Laplace noise samples.
 */
fun interface NoiseGenerator {
  fun sample(scale: Double): Double
}

class LaplaceNoiseGenerator(private val random: Random = Random.Default) : NoiseGenerator {
  override fun sample(scale: Double): Double {
    if (scale == 0.0) return 0.0
    val u = random.nextDouble() - 0.5
    val magnitude = ln(1 - 2 * abs(u)) * -scale
    return sign(u) * magnitude
  }
}

/**
 * Applies per-metric Laplace noise prior to upload.
 */
class DifferentialPrivacyEngine(
  private val noiseGenerator: NoiseGenerator = LaplaceNoiseGenerator()
) {
  fun sanitize(metrics: List<Metric>, policyPack: PolicyPack): List<Metric> {
    if (metrics.isEmpty()) return metrics
    return metrics.map { metric ->
      val policy = policyPack.metricPolicyFor(metric.name)
      if (policy == null) {
        metric
      } else if (!policy.enableDifferentialPrivacy) {
        metric
      } else {
        val scale = if (policy.epsilon == 0.0) 0.0 else policy.sensitivity / policy.epsilon
        val noise = noiseGenerator.sample(scale)
        val sanitised = when (policy.type) {
          MetricType.COUNT -> max(0.0, (metric.value + noise)).roundToInt().toDouble()
          MetricType.SUM -> metric.value + noise
        }
        metric.copy(value = sanitised)
      }
    }
  }

  fun errorBound(policy: MetricPolicy, confidence: Double): Double {
    require(confidence in 0.0..1.0) { "confidence must be between 0 and 1" }
    val clamped = confidence.coerceIn(0.0, 0.999999)
    val scale = if (policy.epsilon == 0.0) 0.0 else policy.sensitivity / policy.epsilon
    if (scale == 0.0) return 0.0
    return scale * ln(1.0 / (1.0 - clamped))
  }
}

File: sdk/mobile/ios/Sources/ODPS/DifferentialPrivacy.swift
import Foundation

public struct SeededGenerator: RandomNumberGenerator {
    private var state: UInt64

    public init(seed: UInt64) {
        self.state = seed == 0 ? 1 : seed
    }

    public mutating func next() -> UInt64 {
        state = 2862933555777941757 &* state &+ 3037000493
        return state
    }
}

public struct LaplaceNoiseGenerator {
    private var rng: SeededGenerator

    public init(seed: UInt64) {
        self.rng = SeededGenerator(seed: seed)
    }

    public mutating func sample(scale: Double) -> Double {
        guard scale > 0 else { return 0 }
        let uniform = Double.random(in: 0..<1, using: &rng) - 0.5
        let magnitude = log(1 - 2 * abs(uniform)) * -scale
        return (uniform >= 0 ? 1.0 : -1.0) * magnitude
    }
}

public final class DifferentialPrivacyEngine {
    private var noiseGenerator: LaplaceNoiseGenerator

    public init(noiseGenerator: LaplaceNoiseGenerator) {
        self.noiseGenerator = noiseGenerator
    }

    public convenience init(seed: UInt64 = 1) {
        self.init(noiseGenerator: LaplaceNoiseGenerator(seed: seed))
    }

    public func sanitize(metrics: [Metric], pack: PolicyPack) -> [Metric] {
        guard !metrics.isEmpty else { return metrics }
        var sanitized: [Metric] = []
        var generator = noiseGenerator

        for metric in metrics {
            guard let policy = pack.metricPolicy(for: metric.name) else {
                sanitized.append(metric)
                continue
            }
            guard policy.enableDifferentialPrivacy else {
                sanitized.append(metric)
                continue
            }
            let scale = policy.sensitivity / policy.epsilon
            var localGenerator = generator
            let noise = localGenerator.sample(scale: scale)
            generator = localGenerator

            switch policy.type {
            case .count:
                let noisy = max(0, Int(round(metric.value + noise)))
                sanitized.append(Metric(name: metric.name, type: .count, value: Double(noisy)))
            case .sum:
                sanitized.append(Metric(name: metric.name, type: .sum, value: metric.value + noise))
            }
        }
        noiseGenerator = generator
        return sanitized
    }

    public func errorBound(for policy: MetricPolicy, confidence: Double) -> Double {
        precondition(confidence >= 0 && confidence <= 1, "confidence must be in [0,1]")
        if policy.epsilon == 0 { return 0 }
        let scale = policy.sensitivity / policy.epsilon
        if scale == 0 { return 0 }
        let clamped = min(confidence, 0.999999)
        return scale * log(1 / (1 - clamped))
    }
}
